### PR TITLE
config: de-templatize source/common/config

### PR DESCRIPTION
**Pull request by fredlas**
A refactor to remove a bunch of templating from the xDS implementation, all the way down to the Subscription and SubscriptionCallbacks interfaces. Until now, the subscription implementations have all been templated on resource proto type, but that template type is really only used to convert Any protobuf blobs to the appropriate type before handing them off to the RDS, CDS, etc code. This PR has those Any blobs remain unconverted until they reach e.g. rds_impl.cc, which knows to MessageUtil::anyConvert<RDS>() them.

Removing understanding of type from the xDS protocol implementation brings things more into line with discovery.proto, which view xDS resources as Any blobs, and generally knows nothing about resource type.

Contributes to #5270.

Risk Level: low
Testing: existing tests

### CODE REVIEW

**Comment by fredlas** [comment]
All the classes that were detemplatized should be split out into .cc files, but I figured I would save that until after the initial review to make the diff easier to review.

Motivation: my aggregated delta implementation takes the approach of a "context" (playing a similar role to GrpcMux), which has a map of type_url -> DeltaSubscriptionState, where DeltaSubscriptionState is the logic that tracks our interest in+version of resources, builds request protos, and handles response protos after they are received. Here is a quick WIP snapshot of that work:
https://github.com/fredlas/envoy/blob/ADS_incremental/source/common/config/delta_subscription_impl.h
https://github.com/fredlas/envoy/blob/ADS_incremental/source/common/config/grpc_delta_xds_context.hhttps://github.com/fredlas/envoy/blob/ADS_incremental/source/common/config/delta_subscription_state.h

File: source/common/config/subscription_factory.h [hunk 1]
```diff
@@ -37,19 +37,18 @@ class SubscriptionFactory {
...
    */
-  template <class ResourceType>
-  static std::unique_ptr<Subscription<ResourceType>> subscriptionFromConfigSource(
+  static std::unique_ptr<Subscription> subscriptionFromConfigSource(
       const envoy::api::v2::core::ConfigSource& config, const LocalInfo::LocalInfo& local_info,
       Event::Dispatcher& dispatcher, Upstream::ClusterManager& cm, Runtime::RandomGenerator& random,
       Stats::Scope& scope, const std::string& rest_method, const std::string& grpc_method,
-      Api::Api& api) {
-    std::unique_ptr<Subscription<ResourceType>> result;
+      const std::string& type_url, Api::Api& api) {
```
Comment (htuch) [inline source/common/config/subscription_factory.h]: Nit: why not `string_view`? You seem to be making that improvement elsewhere :)

File: test/common/upstream/eds_test.cc [hunk 2]
```diff
@@ -30,7 +30,10 @@ namespace {
 
 class EdsTest : public testing::Test {
 protected:
-  EdsTest() : api_(Api::createApiForTest(stats_)) { resetCluster(); }
+  EdsTest() : api_(Api::createApiForTest(stats_)) {
+    resetCluster();
+    fare_cluster_load_assignment_.set_cluster_name("fare");
```
Comment (htuch) [inline test/common/upstream/eds_test.cc]: Why this change?

File: test/common/upstream/eds_test.cc [hunk 3]
```diff
@@ -497,8 +507,9 @@ TEST_F(EdsTest, EndpointRemoval) {
 
   add_endpoint(80);
   add_endpoint(81);
-
-  VERBOSE_EXPECT_NO_THROW(cluster_->onConfigUpdate(resources, ""));
+  resources_.Clear();
```
Comment (htuch) [inline test/common/upstream/eds_test.cc]: Nit: I prefer just using brace scoped variables for resources rather than globals that we keep adding/mutating; unless there is some other reason we need them to be globals (e.g. mock setup etc)

File: source/common/config/subscription_factory.h (hunk 1 above)
Comment (fredlas) [reply source/common/config/subscription_factory.h]: Thanks, good catch.

File: test/common/upstream/eds_test.cc (hunk 2 above)
Comment (fredlas) [reply test/common/upstream/eds_test.cc]: Most of the tests use this same "fare" one, and declaring locally uses an extra 3 lines. I thought it was borderline even when I first changed it, so I'm fine with keeping them local.

File: test/common/upstream/eds_test.cc (hunk 3 above)
Comment (fredlas) [reply test/common/upstream/eds_test.cc]: Yeah, definitely, what I had was fairly gross. I think I have a pretty clean approach now.
